File: java/src/com/v2soft/styxlib/library/server/ClientsHandler.java
package com.v2soft.styxlib.library.server;

import java.io.Closeable;
import java.io.IOException;
import java.nio.channels.SocketChannel;
import java.util.HashMap;
import java.util.Map;
import com.v2soft.styxlib.library.server.vfs.IVirtualStyxDirectory;

/**
 * 
 * @author V.Shcriyabets (vshcryabets@gmail.com)
 *
 */
public class ClientsHandler 
    implements Closeable {
    private String mProtocol;
    private int mIOUnit;
    private Map<SocketChannel, ClientState> mClientStatesMap;
    private IVirtualStyxDirectory mRoot;
    
    public ClientsHandler(int iounit, IVirtualStyxDirectory root, String protocol) throws IOException {
        mIOUnit = iounit;
        mClientStatesMap = new HashMap<SocketChannel, ClientState>();
        mRoot = root;
        mProtocol = protocol;
    }

    public void addClient(SocketChannel client) throws IOException {
        client.configureBlocking(false);
        mClientStatesMap.put(client, new ClientState(mIOUnit, client, mRoot, mProtocol));
    }

    @Override
    public void close() throws IOException {
    }

	protected boolean readClient(SocketChannel channel) throws IOException {
		final ClientState state = mClientStatesMap.get(channel);
		boolean result = state.readSocket();
		if ( result ) {
		    removeClient(channel);
		}
		return result;
	}

	
	private void removeClient(SocketChannel channel) throws IOException {
	    final ClientState state = mClientStatesMap.get(channel);
	    mRoot.onConnectionClosed(state);
    	mClientStatesMap.remove(channel);
    	channel.close();
	}
    
    
}

File: native/V2StyxLib/src/ClientsHandler.cpp
/*
 * ClientsHandler.cpp
 *
 *  Created on: May 22, 2012
 *      Author: V.Shcriyabets (vshcryabets@gmail.com)
 */

#include "ClientsHandler.h"
#include "ClientState.h"

ClientsHandler::ClientsHandler(int iounit,
		IVirtualStyxDirectory *root, std::string *protocol) : mIOUnit(iounit), mRoot(root), mProtocol(protocol) {
	mClientStatesMap = new std::map<Socket, ClientState*>();
}

ClientsHandler::~ClientsHandler() {
	delete mClientStatesMap;
}

void ClientsHandler::addClient(Socket client) {
	mClientStatesMap->insert(
			std::pair<Socket, ClientState*>(client,
					new ClientState(mIOUnit, client, mRoot, mProtocol)));
}

bool ClientsHandler::readClient(Socket socket) {
	ClientState *state = mClientStatesMap->find(socket)->second;
	bool result = state->readSocket();
	if ( result ) {
		removeClient(socket);
	}
	return result;
}

void ClientsHandler::removeClient(Socket socket) {
	mClientStatesMap->erase(socket);
	close(socket);
}
